src/algorithms/preprocessing/reweighing.jl

```diff
@@ -0,0 +1,110 @@
+"""
+Helper function for ReweighingWrapper and ReweighingSamplingWrapper.
+grps is an array of values of protected attribute. y is an array of ground truth values.
+Array of (Frequency) weights are returned from this function.
+"""
+function _calculateWeights(grps, y)
+    grps_classes = levels(grps)
+    n_classes = length(grps_classes)
+    n = length(y)
+
+    fav = y.==levels(y)[2] # Boolean
+    unfav = y.==levels(y)[1]
+    n₊ = sum(fav) # Total Number of favourable outcomes
+
+    weights = zeros(Float64, n)
+    for i in 1:n_classes
+        class = grps.==grps_classes[i]
+        weights[class .& fav] .= n₊*sum(class)/(n*sum(class .& fav))
+        weights[class .& unfav] .= (n-n₊)*sum(class)/(n*sum(class .& unfav))
+    end
+    return weights
+end
+
+"""
+    ReweighingWrapper
+
+ReweighingWrapper is a preprocessing algorithm wrapper in which Weights for each group-label combination is calculated.
+These calculated weights are then passed to the classifier model which further uses it to make training fair.
+"""
+mutable struct ReweighingWrapper <: DeterministicNetwork
+    grp::Symbol
+    classifier::MLJBase.Model
+end
+
+"""
+    ReweighingWrapper(classifier; grp=:class)
+
+Instantiates a ReweighingWrapper which wrapper the `classifier` with the Reweighing fairness algorithm.
+The sensitive attribute can be specified by the parameter `grp`.
+If `classifier` doesn't support weights while training, an error is thrown.
+"""
+function ReweighingWrapper(classifier::MLJBase.Model; grp::Symbol=:class)
+    supports_weights(classifier) || throw(ArgumentError("Classifier provided does not support weights"))
+    return ReweighingWrapper(grp, classifier)
+end
+
+function MLJBase.fit(model::ReweighingWrapper,
+    verbosity::Int, X, y)
+
+    Xs = source(X)
+    ys = source(y, kind=:target)
+
+    stand_model = Standardizer()
+    stand = machine(stand_model, Xs)
+    W = transform(stand, Xs)
+
+    grps = X[:, model.grp]
+
+    weights = _calculateWeights(grps, y)
+
+    classifier = model.classifier
+    cls_mch = machine(classifier, W, ys, weights)
+    ŷ= MMI.predict(cls_mch, W)
+
```

Comment:
As of MLJBase 0.12 the magic for doing this is a bit different:

```julia
mach = machine(Deterministic(), Xs, ys; predict=ŷ)
fit!(mach, verbosity=verbosity)
return mach()
```

See https://alan-turing-institute.github.io/MLJ.jl/dev/composing_models/#Method-II:-Finer-control-(advanced)-1 for more details.